content/blog/2019-11-02-custom-alloc/index.md

```diff
@@ -0,0 +1,206 @@
+
+
+**[2] Regions**.  An allocator that allocates large chunks of memory, puts smaller pieces within it, and then must free them all at once. They are fast and convenent, but use a lot of space and are arguably dangrous because dangling references keep things from being freed. This (nominally) requires programmers to re-structure code to keep references to the region, and free the entire region at once, resulting in a usage pattern that looks somewhat like this:
+
+	```
+	createRegion(r);
+	x1 = regionMalloc(r,8);
+	x2 = regionMalloc(r,8);
+	x3 = regionMalloc(r,16);
+	x4 = regionMalloc(r,8);
+	```
+	
+**[3] Custom patterns**. Anything else --- for example, those that exploit stack-like patterns in memory allocation (the relevant benchmark is `197.parser`). The authors describe these as fast, but brittle.
+
+## Experimental Setup
+
+Although this paper is well-regarded and remembered primarily for its evaluation, the actual experimental data involves only 8 programs, each run with a single input. The table containing the programs and their inputs can be seen in the figure below:
+
+<!-- ![](expts.png) -->
+<img src="expts.png" alt="drawing" width="400" style="margin-left:100px;"/>
+
+All of the other graphs in this blog post will reference the 8 benchmarks on the left, and run their experiments for a single input, a single time, not reporting variance, the impact of parameters, machine loads, or any other confounders. For a paper that is purported to be an exemplar takedown of a common design practice, and appears to have had a large impact on the state of memory allocation, its actual empirical backing is quite flimsy.
+
+Given that there are only 8 data points, with no reported statistics, and more than 8 graphs presented in the paper alone (we must imagine the authors conducted far more that they did not report), that dozens of features were exposed, and that the authors were willing to consider non-linear decision boundaries (the _is it a region_ decision stump), makes things even more worrying. The authors also give very little attention to allocators that are not regions. As we will see, the hypothesis that the primary reason for the difference in performance is the existence of regions also does not even fit the data particularly well: many of the region allocators, too, fail to outperform `dlmalloc` by wide margins.
+
+### Emulation
+
+In order to compare allocators which have different semantics and expose different interfaces to the programmers, Emery et. al. write an emulator so that these programs can be run using malloc and free. This allows them to be linked to the analysis tools, and for various properties of the programs to be computed, such as the drag. This hampers the custom allocators and could also be a source of preformance disparity.
```

Comment:
"Emery et. al.": Two things:

- It's usually spelled "et al." (only one period)
- It's "Berger et al." (that's his last name)

(Please do a find/replace; this occurs again below.)